lib/Mailer/SparkPost.php

```diff
@@ -0,0 +1,131 @@
+                'text' => $message->getText(),
+                'html' => $message->getHtml(),
+            ],
+        ];
+
+        if ($message->getReplyTo()) {
+            $parameters['content']['reply_to'] = $message->getReplyTo();
+        }
+
+        if (count($message->getHeaders())) {
+            $parameters['content']['headers'] = $message->getHeaders();
+        }
+
+        if ($message instanceof AttachmentsAwareInterface) {
+            foreach ($message->getAttachments() as $attachment) {
+                $inline = $attachment->getId() !== null;
+                $parameters['content'][$inline ? 'inline_images' : 'attachments'][] = [
+                    'type' => $attachment->getType(),
+                    'name' => $inline ? $attachment->getId() : $attachment->getName(),
+                    'data' => base64_encode($this->getAttachmentContent($attachment)),
+                ];
+            }
+        }
+
+        if ($message instanceof MetadataAwareInterface) {
+            $metadata = array_filter($message->getMetadata());
+            if (count($metadata)) {
+                $parameters['metadata'] = $metadata;
+            }
+        }
+
+        $tags = [];
+        if ($message instanceof TaggableInterface) {
+            $tags = (array) $message->getTag();
+        }
+
+        $normalizedTo = $this->normalizeIdentities($message->getTo());
+        $normalizedCc = $this->normalizeIdentities($message->getCc());
+        $normalizedBcc = $this->normalizeIdentities($message->getBcc());
+        $toIdentityString = IdentityUtils::buildIdentityString($normalizedTo);
+
+        /** @var \Stampie\Identity[] $merged */
+        $merged = array_merge($normalizedTo, $normalizedCc, $normalizedBcc);
+        foreach ($merged as $recipient) {
+            /*
+             * IMPORTANT:
+             * Do not set address.name!
+             * SparkPost will wrap the entire header_to value, e.g. Bob <bob@example.com, john@example.com>
+             */
+
+            $parameters['recipients'][] = [
+                'address' => [
+                    'email' => $recipient->getEmail(),
+                    'header_to' => $toIdentityString,
+                ],
+                'tags' => $tags,
+            ];
+        }
+
+        if (count($normalizedCc)) {
+            $parameters['content']['headers']['Cc'] = IdentityUtils::buildIdentityString($normalizedCc);
+        }
+
+        return json_encode($parameters);
+    }
+
+    protected function handle(ResponseInterface $response)
+    {
+        $httpException = new HttpException($response->getStatusCode(), $response->getStatusText());
+
+        // 4xx will contain error information in the body encoded as JSON
+        if (!in_array($response->getStatusCode(), range(400, 429), false)) {
+            throw $httpException;
+        }
+
+        throw new ApiException($response->getContent(), $httpException);
+    }
+
+    protected function getAttachmentContent(Attachment $attachment)
```

Comment:
should be private

Reply:
I mimicked the Postmark and Mandrill mailers here. Making it private breaks the test at https://github.com/Stampie/Stampie/pull/83/files#diff-7404c8f30d9034dc95f40bd25167575dR183, but I'll try to figure out how to get around it. Might be easiest just to create dummy fixture files. Ideas?

Reply:
creating a dummy file is better than writing tests against a partial mock (as it means part of the code is untested).
Rewriting other tests is actually on my TODO list